source/extensions/common/wasm/wasm.h

```diff
@@ -149,15 +176,39 @@ class PluginHandle : public PluginHandleBase {
 
 using PluginHandleSharedPtr = std::shared_ptr<PluginHandle>;
 
-class PluginHandleSharedPtrThreadLocal : public ThreadLocal::ThreadLocalObject {
+// PluginHandleManager is responsible for recreating a plugin instance in case of VM failures for
+// the same plugin configuration and base_wasm. tryRestartPlugin can be  used to restart at
+// callsites where plugin's lifecycle varies.
+class PluginHandleManager : public ThreadLocal::ThreadLocalObject,
+                            Logger::Loggable<Logger::Id::wasm> {
```

Comment:
Could `PluginHandleManager` be implemented in the Host repo instead?

Reply:
I'm thinking about the feasibility of this.. If we want to achieve this, we also have to put the `bool restartAllowed()` into the C++ host since we have to define it in (base Wasm's) WasmHandle in order to rate limit recovery restarts across multiple plugins sharing a VM. 

But if we change the logic here to "rate limit per plugin restarts", then we could put the all the logics into PluginHandleManager. (I think this is better simplicity-wise but a little bit confusing and counter-intuitive because we put the restart configuration under "vm_config"). wdyt?